### ISSUE TITLE
bindings generated are not FFI-safe when std::string is returned by value

### ISSUE THREAD

**Issue by ian-p-cooke** [rust-lang/rust-bindgen#2071]
### Input C/C++ Header

```C++
// bindgen-flags: --opaque-type "std::.+" --enable-cxx-namespaces --allowlist-type "StdStringReturnedByValue"

namespace std {
template <typename> class allocator;
template <class> struct char_traits;
template <typename a, typename = char_traits<a>, typename = allocator<a>>
class basic_string;
typedef basic_string<char> string;
template <typename> class allocator {};
template <typename, typename, typename> class basic_string {
public:
  struct {};
  long b;
  enum { c = 15 };
  char d[c];
  basic_string(const char *, const allocator<char> & = allocator<char>());
};
} // namespace std

struct StdStringReturnedByValue {
  std::string value();
};
std::string StdStringReturnedByValue::value() {
  std::string value("hello");
  return value;
}
```

### Bindgen Invocation
```
$ bindgen input.h --opaque-type "std::.+" --enable-cxx-namespaces --allowlist-type "StdStringReturnedByValue"
```

### Actual Results
```rust
#![allow(
    dead_code,
    non_snake_case,
    non_camel_case_types,
    non_upper_case_globals
)]

#[allow(non_snake_case, non_camel_case_types, non_upper_case_globals)]
pub mod root {
    #[allow(unused_imports)]
    use self::super::root;
    pub mod std {
        #[allow(unused_imports)]
        use self::super::super::root;
        pub type string = [u64; 4usize];
    }
    #[repr(C)]
    #[derive(Debug, Default, Copy, Clone)]
    pub struct StdStringReturnedByValue {
        pub _address: u8,
    }
    #[test]
    fn bindgen_test_layout_StdStringReturnedByValue() {
        assert_eq!(
            ::std::mem::size_of::<StdStringReturnedByValue>(),
            1usize,
            concat!("Size of: ", stringify!(StdStringReturnedByValue))
        );
        assert_eq!(
            ::std::mem::align_of::<StdStringReturnedByValue>(),
            1usize,
            concat!("Alignment of ", stringify!(StdStringReturnedByValue))
        );
    }
    extern "C" {
        #[link_name = "\u{1}_ZN24StdStringReturnedByValue5valueEv"]
        pub fn StdStringReturnedByValue_value(
            this: *mut root::StdStringReturnedByValue,
        ) -> root::std::string;
    }
    impl StdStringReturnedByValue {
        #[inline]
        pub unsafe fn value(&mut self) -> root::std::string {
            StdStringReturnedByValue_value(self)
        }
    }
}
```

```
warning: `extern` block uses type `[u64; 4]`, which is not FFI-safe
  --> /tmp/bindings.rs.aawdvy:41:14
   |
41 |         ) -> root::std::string;
   |              ^^^^^^^^^^^^^^^^^ not FFI-safe
   |
   = note: `#[warn(improper_ctypes)]` on by default
   = help: consider passing a pointer to the array
   = note: passing raw arrays by value is not FFI-safe
```

### Expected Results
I expected no warnings to be emitted.  tests pass fine but I would like the bindings generated to be FFI-safe.

Is there another set of flags that would do that for me?

In my actual project I'm wrapping an external library that returns std::string sometimes and I get these warnings there.  I worked around it by replacing std::string with another struct with the same size and alignment but I don't like that solution.
```C++
/**
 * <div rustbindgen="true" replaces="std::string"></div>
 */
struct basic_string_char {
    uint64_t storage[4];
};
```

**Comment by ian-p-cooke** [comment]
I set up a test in https://github.com/ian-p-cooke/rust-bindgen/tree/no_warn_on_return_std_string and ran
```
./tests/test-one.sh no_warn > test_one_no_warn.txt 2>&1
```
output is attached.
[test_one_no_warn.txt](https://github.com/rust-lang/rust-bindgen/files/6787721/test_one_no_warn.txt)

**Comment by emilio** [comment]
This is expected, because Rust doesn't implement the C++ ABI with regard to destructors / drop implementations unfortunately. So from the point of view of Rust all types are POD, but C++ changes the ABI for non-trivial types.

**Comment by ian-p-cooke** [comment]
> This is expected, because Rust doesn't implement the C++ ABI with regard to destructors / drop implementations unfortunately. So from the point of view of Rust all types are POD, but C++ changes the ABI for non-trivial types.

can you explain this a bit more?  Does that mean if I called a C++ function that returned std::string by value the memory for the string would be leaked because a destructor would not be called?

It's ok if that's the case, I don't need to call any such functions, I just want to understand the implications in general here.

**Comment by emilio** [comment]
No, there are various issues here. Let's assume Linux x86-64 for simplicity, the details on Windows are more complicated, and other platforms have different calling conventions.

Consider you return a small object that has a destructor like `std::unique_ptr`:

```c
extern "C" std::unique_ptr<Foo> get_foo();
```

On C++, that'd go through https://itanium-cxx-abi.github.io/cxx-abi/abi.html#non-trivial-return-values, while the Rust ABI doesn't have that kind of concept and that'd just go through the regular C ABI and expect the returned pointer in a register as per https://itanium-cxx-abi.github.io/cxx-abi/abi.html#non-trivial.

So small types with destructors have this issue. In general, it's not safe to pass around / return types with destructors by value because Rust doesn't implement the C++ ABI, just the C ABI.

The other main issue is that Rust has this assumption that moving an object can just be done via `memmove()`. As such, types that contain interior pointers like `std::string` may end up having an internal dangling pointer to e.g. another part of the stack where that instance used to be, which is obviously bad.

**Comment by ian-p-cooke** [comment]
ok, I think I understand the issue now.  

I guess what I would like is some way to indicate that I don't want the bindings generated to include funtions/methods that are known to be broken because of C++ ABI issues.  I don't want to explain to my team that it's safe to ignore these warnings because we don't use the methods... I'd rather indicate at build-time that 'these are known to be bad'.

It seems like 'opaque' isn't opaque enough and block-listing std::string leads to incomplete definitions.  I'm ok with manually (explicit list, regex, whatever) indicating which types those are but I can't find a way to do that.

**Comment by emilio** [comment]
Yeah I don't think we have anything like that right now. You can of course blocklist the functions themselves though, and that'd avoid the warnings I suppose?
